sdk/identity/azure-identity/src/client_assertion_credential.cpp

```diff
@@ -0,0 +1,160 @@
+using Azure::Core::Http::HttpMethod;
+using Azure::Identity::_detail::IdentityLog;
+using Azure::Identity::_detail::TenantIdResolver;
+using Azure::Identity::_detail::TokenCredentialImpl;
+
+namespace {
+bool IsValidTenantId(std::string const& tenantId)
+{
+  const std::string allowedChars = ".-";
+  if (tenantId.empty())
+  {
+    return false;
+  }
+  for (auto const c : tenantId)
+  {
+    if (allowedChars.find(c) != std::string::npos)
+    {
+      continue;
+    }
+    if (!StringExtensions::IsAlphaNumeric(c))
+    {
+      return false;
+    }
+  }
+  return true;
+}
+} // namespace
+
+ClientAssertionCredential::ClientAssertionCredential(
+    std::string tenantId,
+    std::string clientId,
+    std::function<std::string(Context const&)> const& assertionCallback,
+    ClientAssertionCredentialOptions const& options)
+    : TokenCredential("ClientAssertionCredential"), m_assertionCallback(assertionCallback),
+      m_clientCredentialCore(tenantId, options.AuthorityHost, options.AdditionallyAllowedTenants)
+{
+  bool isTenantIdValid = IsValidTenantId(tenantId);
+  if (!isTenantIdValid)
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Warning,
+        "Invalid tenant ID provided  for " + GetCredentialName()
+            + ". The tenant ID must be a non-empty string containing only alphanumeric characters, "
+              "periods, or hyphens. You can locate your tenant ID by following the instructions "
+              "listed here: https://learn.microsoft.com/partner-center/find-ids-and-domain-names");
+  }
+  if (clientId.empty())
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Warning, "No client ID specified for " + GetCredentialName() + ".");
+  }
+  if (!assertionCallback)
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Warning,
+        "The assertionCallback must be a valid function that returns assertions for "
+            + GetCredentialName() + ".");
+  }
+
+  if (isTenantIdValid && !clientId.empty() && assertionCallback)
+  {
+    m_tokenCredentialImpl = std::make_unique<TokenCredentialImpl>(options);
+    m_requestBody
+        = std::string(
+              "grant_type=client_credentials"
+              "&client_assertion_type="
+              "urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer" // cspell:disable-line
+              "&client_id=")
+        + Url::Encode(clientId);
+
+    IdentityLog::Write(
+        IdentityLog::Level::Informational, GetCredentialName() + " was created successfully.");
+  }
+  else
+  {
+    // Rather than throwing an exception in the ctor, following the pattern in existing credentials
+    // to log the errors, and defer throwing an exception to the first call of GetToken(). This is
+    // primarily needed for credentials that are part of the DefaultAzureCredential, which this
+    // credential is not intended for.
+    IdentityLog::Write(
+        IdentityLog::Level::Warning, GetCredentialName() + " was not initialized correctly.");
```

Comment:
That will require storing the errors in the class (even in cases when things eventually succeed) and result in duplicated logs being printed. To me, that's more noise than signal since the info would be redundantly present.